worlds/sc2/__init__.py

```diff
@@ -641,15 +641,24 @@ def flag_and_add_resource_locations(world: SC2World, item_list: List[FilterItem]
                 world.locked_locations.append(location.name)
 
 
-def flag_mission_order_required_items(world: SC2World, item_list: List[FilterItem]) -> None:
-    """Marks items that are necessary for item rules in the mission order and forces them to be progression."""
+def create_and_flag_mission_order_required_items(world: SC2World, item_list: List[FilterItem]) -> None:
+    """Marks items that are necessary for item rules in the mission order and forces them to be progression,
+    and create extra items if the existing amount is insufficient."""
     locks_required = world.custom_mission_order.get_items_to_lock()
     locks_done = {item: 0 for item in locks_required}
     for item in item_list:
         if item.name in locks_required and locks_done[item.name] < locks_required[item.name]:
             item.flags |= ItemFilterFlags.Necessary
             item.flags |= ItemFilterFlags.ForceProgression
             locks_done[item.name] += 1
+    # Create extra items if there's not enough in the pool already
+    for (item_name, amount) in locks_required.items():
+        done = locks_done[item_name]
+        for idx in range(amount - done):
+            new_item = FilterItem(item_name, items.item_table[item_name], done + idx)
+            new_item.flags |= ItemFilterFlags.Necessary
+            new_item.flags |= ItemFilterFlags.ForceProgression
+            item_list.append(new_item)
```

Comment:
Two things:
* This doesn't account for the case where keys adds more copies of an item than it's `quantity` field _and_ the user has a higher-than-quantity starting inventory amount. I recommend moving this logic to `create_and_flag_explicit_item_locks_and_excludes()`, and just using the the key list / `get_items_to_lock()` to possibly increase the maximum item count if it's greater than `item.quantity`. Can be done with just a max()
* I can't really see a difference between necessary and locked items. I recommend just using locked

Reply:
First point is fair, second is wrong. The original Item rule PR ran into this, but locked items are allowed to be deleted from the item pool during pool filtering if they're entirely useless, and there's some special cases that make it difficult to work around. I had a conversation about this with Zik, and the outcome was that a necessary flag is just better.

Reply:
Could you add a comment on the Necessary definition to specify, then? I saw this, was confused, and searched around the code for a good 5 minutes and couldn't spot any issues, so it's better to save future readers the time to just say what the difference is up-front.

I still think it's better to consolidate them and just make locked behave like what you're describing necessary as, but I'm fine to call that out-of-scope for this PR